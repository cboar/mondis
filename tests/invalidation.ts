import type { CollectedInvalidations } from '../src/CachedQuery/invalidation/core';
import { mondisTest, oid, type ModelMap } from './setup';

type UpdateMap = Record<string, ((map: ModelMap) => Promise<unknown>)>;

const MongooseUpdates: UpdateMap = {
  InsertCar: ({ Vehicle }) => (
    Vehicle.create({ _id: oid('A5'), kind: 'car', price: 6000, routes: [] })
  ),
  InsertTruckWithDriver: ({ Vehicle }) => (
    Vehicle.create({ _id: oid('B5'), kind: 'truck', driver: oid('D2'), price: 10000, routes: ['C'] })
  ),
  InsertDriver: ({ Driver }) => (
    Driver.create({ _id: oid('D5'), name: 'Frank', salary: 51000 })
  ),

  UpdateCarPrice: ({ Vehicle }) => (
    Vehicle.updateOne({ kind: 'car' }, { $inc: { price: 8000 } }).exec()
  ),
  UpdateAllPrices: ({ Vehicle }) => (
    Vehicle.updateMany({}, { $inc: { price: 2000 } }).exec()
  ),
  UpdateCarDriver: ({ Vehicle }) => (
    Vehicle.updateOne({ kind: 'car', driver: oid('D1') }, { driver: oid('D2') }).exec()
  ),
  UpdateCarToBus: ({ Vehicle }) => (
    Vehicle.updateOne({ kind: 'car' }, { kind: 'bus' }).exec()
  ),
  UpdateHenrySalary: ({ Driver }) => (
    Driver.updateOne({ name: 'Henry' }, { salary: 60000 }).exec()
  ),
  UpdateJohnName: ({ Driver }) => (
    Driver.updateOne({ name: 'John' }, { name: 'Jon' }).exec()
  ),
  UpdateBusRoutes: ({ Vehicle }) => (
    Vehicle.updateOne({ kind: 'bus' }, { $push: { routes: 'B' } }).exec()
  ),
  UpsertCarViaPrice: ({ Vehicle }) => (
    Vehicle.updateOne({ kind: 'car', price: 11000 }, { price: 15000 }, { upsert: true }).exec()
  ),

  RemoveCar: ({ Vehicle }) => (
    Vehicle.deleteOne({ kind: 'car' }).exec()
  ),
  RemoveTruck: ({ Vehicle }) => (
    Vehicle.deleteOne({ kind: 'truck' }).exec()
  ),
  RemoveHenry: ({ Driver }) => (
    Driver.deleteOne({ name: 'Henry' }).exec()
  ),
};

Object.entries(MongooseUpdates).forEach(([name, execUpdate]) => {
  mondisTest(name, async ({ mondis, models }) => {
    jest.spyOn(mondis.invalidator, 'doInvalidations').mockImplementation((collected: CollectedInvalidations) => {
      collected.keys?.sort((a, b) => (a > b ? 1 : -1));
      collected.sets?.sort((a, b) => (a > b ? 1 : -1));
      expect(collected).toMatchSnapshot();
      return Promise.resolve();
    });
    await execUpdate(models);
  });
});

// TODO: test to ensure keys are actually invalidated.

// TODO: utils unit tests
// testCase({ inclusive: false, paths: [] }, [], false);
// testCase({ inclusive: false, paths: [] }, ['lol'], true);
// testCase({ inclusive: true, paths: ['hello'] }, ['hello.world'], true);
// testCase({ inclusive: true, paths: ['hello.world'] }, ['hello.world'], true);
// testCase({ inclusive: true, paths: ['hello.world'] }, ['hello.lol'], false);
// testCase({ inclusive: true, paths: ['hello.world'] }, ['hello'], true);
// testCase({ inclusive: true, paths: ['hello', 'lol'] }, ['lol'], true);
// testCase({ inclusive: true, paths: ['hello', 'lol'] }, ['nah'], false);
// testCase({ inclusive: false, paths: ['hello'] }, ['hello.world'], false);
// testCase({ inclusive: false, paths: ['hello.world'] }, ['hello.world'], false);
// testCase({ inclusive: false, paths: ['hello.world'] }, ['hello.lol'], true);
// testCase({ inclusive: false, paths: ['hello', 'lol'] }, ['hello', 'lol'], false);
// testCase({ inclusive: false, paths: ['hello', 'lol'] }, ['hello', 'lol', 'cool'], true);
// testCase({ inclusive: false, paths: ['hello.world'] }, ['hello'], true);
// testCase({ inclusive: false, paths: ['hello.world'] }, ['hello.world.ok'], false);
