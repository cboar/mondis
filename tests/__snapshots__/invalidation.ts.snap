// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsertCar 1`] = `
{
  "keys": [
    "Q:Dynamic1________["car"]",
    "Q:Dynamic3________[null]",
    "Q:Populated1______[]",
    "Q:Populated2______[]",
    "Q:Static1_________[]",
  ],
  "sets": [
    {
      "set": "A:Complex2________",
    },
    {
      "set": "A:Complex1________",
    },
    {
      "set": "A:Dynamic2________",
    },
  ],
}
`;

exports[`InsertDriver 1`] = `
{
  "keys": [],
  "sets": [],
}
`;

exports[`InsertTruckWithDriver 1`] = `
{
  "keys": [
    "Q:Dynamic1________["truck"]",
    "Q:Populated1______[]",
    "Q:Populated2______[]",
    "Q:Sorted1_________[]",
    "Q:Static2_________[]",
  ],
  "sets": [
    {
      "set": "A:Complex2________",
    },
    {
      "set": "A:Complex1________",
    },
    {
      "set": "A:Dynamic2________",
    },
  ],
}
`;

exports[`RemoveCar 1`] = `
{
  "sets": [
    {
      "set": "P:0000000000000000000000a1",
    },
    {
      "set": "O:0000000000000000000000a1",
    },
  ],
}
`;

exports[`RemoveHenry 1`] = `
{
  "sets": [
    {
      "set": "P:0000000000000000000000d1",
    },
    {
      "set": "O:0000000000000000000000d1",
    },
  ],
}
`;

exports[`RemoveTruck 1`] = `
{
  "sets": [
    {
      "set": "P:0000000000000000000000b1",
    },
    {
      "set": "O:0000000000000000000000b1",
    },
  ],
}
`;

exports[`UpdateAllPrices 1`] = `
{
  "keys": [
    "Q:Dynamic1________["bus"]",
    "Q:Dynamic1________["car"]",
    "Q:Dynamic1________["truck"]",
    "Q:Sorted1_________[]",
    "Q:Static2_________[]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000a1"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000a2"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000a3"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000a4"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000b1"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000b2"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000b3"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000b4"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000c1"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000c2"}]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000c3"}]",
  ],
  "sets": [
    {
      "set": "A:Targeted1_______",
    },
    {
      "set": "A:Complex2________",
    },
  ],
}
`;

exports[`UpdateBusRoutes 1`] = `
{
  "keys": [
    "Q:Dynamic3________[null]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000c1"}]",
  ],
  "sets": [
    {
      "set": "A:Targeted1_______",
    },
    {
      "set": "A:Dynamic2________",
    },
  ],
}
`;

exports[`UpdateCarDriver 1`] = `
{
  "keys": [
    "Q:Populated1______[]",
    "Q:Populated2______[]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000a2"}]",
  ],
  "sets": [],
}
`;

exports[`UpdateCarPrice 1`] = `
{
  "keys": [
    "Q:Dynamic1________["car"]",
    "Q:Sorted1_________[]",
    "Q:Static2_________[]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000a1"}]",
  ],
  "sets": [
    {
      "set": "A:Targeted1_______",
    },
    {
      "set": "A:Complex2________",
    },
  ],
}
`;

exports[`UpdateCarToBus 1`] = `
{
  "keys": [
    "Q:Static1_________[]",
    "Q:Unique1_________[{"$oid":"0000000000000000000000a1"}]",
  ],
  "sets": [
    {
      "set": "A:Targeted1_______",
    },
    {
      "set": "A:Complex1________",
    },
  ],
}
`;

exports[`UpdateHenrySalary 1`] = `
{
  "keys": [],
  "sets": [
    {
      "filter": "Q:Populated1______",
      "set": "P:0000000000000000000000d1",
    },
  ],
}
`;

exports[`UpdateJohnName 1`] = `
{
  "keys": [],
  "sets": [
    {
      "filter": "Q:Populated2______",
      "set": "P:0000000000000000000000d2",
    },
    {
      "filter": "Q:Populated1______",
      "set": "P:0000000000000000000000d2",
    },
  ],
}
`;

exports[`UpsertCarViaPrice 1`] = `
{
  "keys": [
    "Q:Dynamic1________["car"]",
    "Q:Dynamic2________[null]",
    "Q:Populated1______[]",
    "Q:Populated2______[]",
    "Q:Sorted1_________[]",
    "Q:Static1_________[]",
    "Q:Static2_________[]",
  ],
  "sets": [
    {
      "set": "A:Complex2________",
    },
    {
      "set": "A:Complex1________",
    },
  ],
}
`;
